core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -187,44 +239,35 @@ private[log] class ProducerAppendInfo(val producerId: Long,
                          offset: Long,
                          timestamp: Long): CompletedTxn = {
     if (isFenced(producerEpoch))
-      throw new ProducerFencedException(s"Invalid producer epoch: $producerEpoch (zombie): ${this.producerEpoch} (current)")
+      throw new ProducerFencedException(s"Invalid producer epoch: $producerEpoch (zombie): ${currentEntry.producerEpoch} (current)")
 
-    if (this.coordinatorEpoch > endTxnMarker.coordinatorEpoch)
+    if (currentEntry.coordinatorEpoch > endTxnMarker.coordinatorEpoch)
       throw new TransactionCoordinatorFencedException(s"Invalid coordinator epoch: ${endTxnMarker.coordinatorEpoch} " +
-        s"(zombie), $coordinatorEpoch (current)")
+        s"(zombie), ${currentEntry.coordinatorEpoch} (current)")
 
-    if (producerEpoch != this.producerEpoch) {
-      // it is possible that this control record is the first record seen from a new epoch (the producer
-      // may fail before sending to the partition or the request itself could fail for some reason). In this
-      // case, we bump the epoch and reset the sequence numbers
-      this.producerEpoch = producerEpoch
-      this.firstSeq = RecordBatch.NO_SEQUENCE
-      this.lastSeq = RecordBatch.NO_SEQUENCE
-    } else {
-      // the control record is the last append to the log, so the last offset will be updated to point to it.
-      // However, the sequence numbers still point to the previous batch, so the duplicate check would no longer
-      // be correct: it would return the wrong offset. To fix this, we treat the control record as a batch
-      // of size 1 which uses the last appended sequence number.
-      this.firstSeq = this.lastSeq
-    }
+    // TODO(reviewers): The semantics of the ProducerIdEntry have changed so that now explicitly caches the metadata
```

Comment:
Actually, the case where the control batch has a new epoch is being handled in `currentEntry.maybeUpdateEpoch`. If the epoch changes we update the epoch in the `ProducerIdEntry` and clear all the batch metadata, which resets all the sequence numbers producer id. 

I am not sure I follow your point about producer id expiration. We expire producer ids by time. However, when we are deleting segments whose retention period is elapsed, we use offsets to determine whether the producer snapshot should go. This could mean that the producer id is removed even though a control message is left. This should be fine, because that message is orphaned, and will be deleted eventually once it violates retention. I don't think this is a correctness issue at all.

Reply:
Yes, I said it made me uncomfortable, not that it was necessarily a problem. I would be fine renaming the method to `lastDataOffset` as suggested above so that it's clear that it is not intended to include control records.

Reply:
We can remove the `TODO` now?